### ISSUE TITLE
Iterating through eccodes.GribFile throws "ValueError: I/O operation on closed file"

### ISSUE THREAD

**Issue by blazk** [ecmwf/eccodes-python#14]
I tried running an example similar to https://confluence.ecmwf.int/display/ECC/High-level+Pythonic+Interface+in+ecCodes
```
#!/usr/bin/env python3

import os, eccodes

input_path = os.path.join(eccodes.codes_samples_path(), 'GRIB2.tmpl')
with eccodes.GribFile(input_path) as grib:
    for msg in grib:
        print(msg['date'])
```
it throws an error:
```
Traceback (most recent call last):
  File "./x", line 7, in <module>
    for msg in grib:
ValueError: I/O operation on closed file
```
This can be fixed by adding [`._next_()`](https://docs.python.org/3/library/stdtypes.html#iterator.__next__) method in [`CodesFile class`](https://github.com/ecmwf/eccodes-python/blob/80be127f1fec4fbefe461b2879d9465f4d1804d7/eccodes/high_level/codesfile.py#L12):
```
    def __next__(self):
        return self.next()
```
Otherwise `._next_()` inherited from the FileIO base class is called and throws "ValueError: I/O operation on closed file".

Maybe fundamentally the problem is `CodesFile.__init__()` not calling the [constructor of its base class (`io.FileIO`)](https://github.com/python/cpython/blob/b08d3f71beab59653edfbbcf7b92a7bc8050d6b8/Modules/_io/fileio.c#L222) and as a result `CodesFile` objects not being [properly initialized](https://github.com/python/cpython/blob/b08d3f71beab59653edfbbcf7b92a7bc8050d6b8/Modules/_io/fileio.c#L385). The fix above enables iteration, but other stuff still doesn't work properly:
```
>>> with eccodes.GribFile('sample.grb') as grib:
...     print(grib.closed)
...     print(grib.fileno())
True
Traceback (most recent call last):
  File "<stdin>", line 3, in <module>
ValueError: I/O operation on closed file
```

**Comment by cofinoa** [comment]
@blazk as I have also discover, this high level API, needs to be refactored. 

I neither understand the `io.FileIO` as base class for `CodesFile`

**Comment by rapto** [comment]
It seems to be fixed in 1.0
